flink-connector-aws/flink-connector-dynamodb/src/main/java/org/apache/flink/connector/dynamodb/source/enumerator/tracker/SplitTracker.java

```diff
@@ -0,0 +1,252 @@
+import static org.apache.flink.connector.dynamodb.source.enumerator.SplitAssignmentStatus.FINISHED;
+
+/** This class is used to track splits and will be used to assign any unassigned splits. */
+@Internal
+public class SplitTracker {
+    private final Map<String, DynamoDbStreamsShardSplit> knownSplits = new ConcurrentHashMap<>();
+    private final Set<String> assignedSplits = new HashSet<>();
+    private final Set<String> finishedSplits = new HashSet<>();
+    private final String streamArn;
+    private final InitialPosition initialPosition;
+
+    public SplitTracker(String streamArn, InitialPosition initialPosition) {
+        this(Collections.emptyList(), streamArn, initialPosition);
+    }
+
+    public SplitTracker(
+            List<DynamoDBStreamsShardSplitWithAssignmentStatus> initialState,
+            String streamArn,
+            DynamodbStreamsSourceConfigConstants.InitialPosition initialPosition) {
+        this.streamArn = streamArn;
+        this.initialPosition = initialPosition;
+        initialState.forEach(
+                splitWithStatus -> {
+                    DynamoDbStreamsShardSplit currentSplit = splitWithStatus.split();
+                    knownSplits.put(currentSplit.splitId(), currentSplit);
+
+                    if (ASSIGNED.equals(splitWithStatus.assignmentStatus())) {
+                        assignedSplits.add(splitWithStatus.split().splitId());
+                    }
+                    if (FINISHED.equals(splitWithStatus.assignmentStatus())) {
+                        finishedSplits.add(splitWithStatus.split().splitId());
+                    }
+                });
+    }
+
+    /**
+     * Add newly discovered splits to tracker.
+     *
+     * @param shardsToAdd collection of splits to add to tracking
+     */
+    public void addSplits(Collection<Shard> shardsToAdd) {
+        Map<String, Shard> shardIdToShardMap = getShardIdToShardMap(shardsToAdd);
+
+        for (Shard shard : shardsToAdd) {
+            String shardId = shard.shardId();
+            if (!knownSplits.containsKey(shardId)) {
+                DynamoDbStreamsShardSplit newSplit =
+                        mapToSplit(shard, getStartingPosition(shard, shardIdToShardMap));
+                knownSplits.put(shardId, newSplit);
+            }
+        }
+    }
+
+    private InitialPosition getStartingPosition(Shard shard, Map<String, Shard> shardIdToShardMap) {
+        if (shard.parentShardId() == null) {
+            return initialPosition;
+        }
+        Shard parentShard = shardIdToShardMap.get(shard.parentShardId());
+        if (parentShard == null) {
+            return TRIM_HORIZON;
+        }
+        InitialPosition parentStartingPosition =
+                getStartingPosition(parentShard, shardIdToShardMap);
+        if (parentStartingPosition == initialPosition) {
+            return TRIM_HORIZON;
+        }
+        return parentStartingPosition;
+    }
```

Comment:
I don't understand why we need to resolve this starting position here. Since we only support `LATEST` and `TRIM_HORIZON`, could we not do the below logic:

1. If `TRIM_HORIZON` configured, every shard will be `TRIM_HORIZON`.
2. If `LATEST` configured, we filter such that we only consume active shards, and every shard will be `LATEST`